### PR TITLE
feat: 2fa

### LINKED ISSUE #124: [Feature]: 2-Factor-Authentication
### Tell us about your Wish

Implement TOTP 2FA

### Is your feature related to a bug

_No response_

### Alternatives

_No response_

### Additional Info

_No response_

### Participation

- [x] I am willing to submit a pull request to implement this

### Disclaimer

- [X] I agree

### PR DESCRIPTION
**Pull request by lennartkloock**
## Proposed changes

Implement 2-Factor-Authentication

- API Endpoints
- Frontend

## Types of changes

What types of changes does your code introduce to Scuffle?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/ScuffleTV/scuffle/blob/main/CONTRIBUTING.md) doc
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

Closes #124

### CODE REVIEW

**Comment by IS2511** [comment]
\> feat: 2fa
\> removes outdated docs
Based?

**Comment by lennartkloock** [comment]
> feat: 2fa > removes outdated docs Based?

Yeah I didn't want to make a new pr just for removing outdated docs

File: platform/website/src/components/auth/solve-two-fa-dialog.svelte [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+						}
+					}
+				`),
+				{ code: code.value },
+				{
+					requestPolicy: "network-only",
+				},
+			)
+			.toPromise()
+			.then((res) => {
```
Comment (EsdrasAmora) [inline platform/website/src/components/auth/solve-two-fa-dialog.svelte:64]: can you use async/await here?

File: platform/website/src/components/settings/account/enable-2fa.svelte [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+		}
+	}
+
+	function next(fromForm: boolean) {
+		if (state.step === 1) {
+			generateSecret();
+		} else if (state.step === 2) {
+			if (fromForm) {
+				enableTotp();
+			}
```
Comment (EsdrasAmora) [inline platform/website/src/components/settings/account/enable-2fa.svelte:30]:
```suggestion
		} else if (state.step === 2 & fromForm) {
			enableTotp();
```

Comment (EsdrasAmora) [reply platform/website/src/components/settings/account/enable-2fa.svelte:30]: There are other if/elses that could be simplified, but I don't want to be too nit.

File: platform/api/src/api/v1/gql/error.rs [hunk 3]
```diff
@@ -16,44 +18,58 @@ impl GqlErrorInterface {
...
+            GqlError::Auth(AuthError::NotLoggedIn) => "Auth(NotLoggedIn)",
+            GqlError::Auth(AuthError::InvalidToken) => "Auth(InvalidToken)",
+            GqlError::Auth(AuthError::UnsolvedTwoFaChallenge) => "Auth(UnsolvedTwoFaChallenge)",
             GqlError::NotImplemented => "NotImplemented",
             GqlError::Unauthorized { .. } => "Unauthorized",
-            GqlError::NotLoggedIn => "NotLoggedIn",
             GqlError::NotFound(_) => "NotFound",
+            GqlError::Totp(_) => "Totp",
         }
     }
```
[the remark this answers is not in the document]
Comment (lennartkloock) [reply platform/api/src/api/v1/gql/error.rs:74]: I'm not sure if it makes sense to introduce a new dependency for a task this simple

File: platform/api/src/api/v1/gql/models/user.rs [hunk 4]
```diff
@@ -42,6 +42,8 @@ pub struct User {
     pub email_verified_: bool,
     #[graphql(skip)]
     pub last_login_at_: DateRFC3339,
+    #[graphql(skip)]
+    pub totp_enabled_: bool,
 }
 
 /// TODO: find a better way to check if a user is allowed to read a field.
```
[the remark this answers is not in the document]
Comment (lennartkloock) [reply platform/api/src/api/v1/gql/models/user.rs:49]: I'm aware that there are field guard. I just hadn't time to refactor the code to use them. It's an active issue though. #131

File: platform/api/src/api/v1/gql/mutations/user/two_fa.rs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+        // Enable 2fa and save backup codes.
+        let user: database::User = sqlx::query_as(
+            "UPDATE users SET totp_enabled = true, updated_at = NOW() WHERE id = $1 RETURNING *",
+        )
+        .bind(auth.session.user_id)
+        .fetch_one(global.db.as_ref())
+        .await?;
+
+        // TODO: Log out all other sessions?
```
[the remark this answers is not in the document]
Comment (lennartkloock) [reply platform/api/src/api/v1/gql/mutations/user/two_fa.rs:126]: I think we won't do it for now

**Review by EsdrasAmora — APPROVED** [review]

**Pull request merged**